### PR TITLE
Add support for Babel configs other than `babel.config.js` in install-expo-modules

### LINKED ISSUE #30355: install-expo-modules crashes if babel.config.json is used instead of babel.config.js
### Summary

I tried to use install-expo-modules to migrate from plain RN to Expo, and I was immediately met with this error:

```
✔ We recommend installing the Expo CLI integration for the best experience. Not using it may result in some features not working as expected. Learn more: https://docs.expo.dev/bare/using-expo-cli/
Install the Expo CLI integration? … yes
› Updating your project...
Uncaught Error Error: [ios.dangerous]: withIosDangerousBaseMod: ENOENT: no such file or directory, open '…/app-native/babel.config.js'
    at async open (node:internal/fs/promises:639:25)
    at async Object.readFile (node:internal/fs/promises:1246:14)
    at async /private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:745010
    at async action (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:560958)
    at async interceptingMod (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:559653)
    at async interceptingMod (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:559653)
    at async interceptingMod (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:559653)
    at async action (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:533863)
    at async interceptingMod (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:559653)
    at async evalModsAsync (/private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-0b5771f5/dlx-72014/node_modules/install-expo-modules/build/index.js:34:541579) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '…/app-native/babel.config.js'
}
```

### What platform(s) does this occur on?

iOS

### SDK Version

n/a

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 22.4.1 - /private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-8c7ce9a9/node
      Yarn: 4.3.1 - /private/var/folders/3y/swkjq1x101b76mzjgt4pm2d80000gp/T/xfs-8c7ce9a9/yarn
      npm: 10.8.1 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      react: 18.2.0 => 18.2.0 
      react-native: ^0.74.1 => 0.74.1 
    Expo Workflow: bare

### Minimal reproducible example

n/a

### PR DESCRIPTION
**Pull request by wojtekmaj**
# Why

Closes #30355

# How

Instead of hardcoding babel.config.js as the only Babel config that possibly can exist and crashing if it doesn't exist, I iterate through an array of possible Babel config files and use whichever is found first. If this still fails, a new warning will be printed out, suggesting manual Babel config change, instead of crashing the migration completely.

# Test Plan

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts [hunk 1]
```diff
@@ -65,10 +65,28 @@ const withCliBabelConfig: ConfigPlugin = (config) => {
...
+        contents = updateBabelConfig(contents);
+        await fs.promises.writeFile(babelConfigJsPath, contents);
+        return config;
+      }
+      
+      const babelConfigJsonPath = path.join(config.modRequest.projectRoot, 'babel.config.json');
+
+      if (await fs.promises.exists(babelConfigJsonPath)) {
+        let contents = await fs.promises.readFile(babelConfigJsonPath, 'utf8');
+        contents = updateBabelConfig(contents);
```
Comment (vonovak) [inline packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts:81]: I'm afraid the result might be invalid json syntax: https://github.com/expo/expo/blob/2de789861daaeb6e658c4d5b2c12415ad61b8b6d/packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts#L116

Comment (wojtekmaj) [reply packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts:81]: Nice catch. It wasn't tested - expo repository was too big to even clone, so I had to write it from GitHub.com.

Comment (wojtekmaj) [reply packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts:81]: I think this should be fine now!

File: packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts [hunk 2]
```diff
@@ -65,10 +65,28 @@ const withCliBabelConfig: ConfigPlugin = (config) => {
   return withDangerousMod(config, [
     'ios',
     async (config) => {
-      const babelConfigPath = path.join(config.modRequest.projectRoot, 'babel.config.js');
-      let contents = await fs.promises.readFile(babelConfigPath, 'utf8');
-      contents = updateBabelConfig(contents);
-      await fs.promises.writeFile(babelConfigPath, contents);
+      const babelConfigJsPath = path.join(config.modRequest.projectRoot, 'babel.config.js');
+
+      if (await fs.promises.exists(babelConfigJsPath)) {
```
Comment (Kudo) [inline packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts:70]: `fs.promises` doesn't have the `exists` function: https://nodejs.org/api/fs.html#promises-api.

**Review by Kudo — COMMENTED** [review]
install-expo-modules is a tool for rnc cli template without deep customization. if we are going to support babel.config.json, then we should further support `.babelrc.json`, `.babelrc.js`, or even babel configuration inside `package.json` https://babeljs.io/docs/configuration.
that's why i didn't have babel.config.json support. what do you think if we keep only babel.config.js or we should further support all babel config variants?

File: packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts (hunk 2 above)
Comment (wojtekmaj) [reply packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts]: Fixed, thank you!

**Comment by wojtekmaj** [comment]
I think migrating Babel config from JS to JSON is fairly common as this poses fewer problems for ESM repositories. We should support all standalone Babel configs (fixed my code to cover more cases). As for the package.json case, perhaps it would be a bit more complicated to add full support for, but at the very least, it's not crashing the CLI now, but provides a helpful warning instead.

File: packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts [hunk 3]
```diff
@@ -65,10 +65,36 @@ const withCliBabelConfig: ConfigPlugin = (config) => {
...
+        }
+
+        let contents = await fs.promises.readFile(babelConfigPath, 'utf8');
+        contents = updateBabelConfig(contents);
+        await fs.promises.writeFile(babelConfigPath, contents);
+        return config;
+      }
+
+      console.warn(
+        '⚠️  Could not find `babel.config.json`, `.babelrc.json`, `babel.config.js`, or `.babelrc.js` in the project root. Please manually update the Babel config to use `babel-preset-expo`.',
```
Comment (wojtekmaj) [inline packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts:96]:
```suggestion
        '⚠️  Could not find Babel config file in the project root. Please manually update the Babel config to use `babel-preset-expo`.',
```

File: packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts [hunk 4]
```diff
@@ -65,10 +65,36 @@ const withCliBabelConfig: ConfigPlugin = (config) => {
...
+        '.babelrc.cjs',
+        '.babelrc.mjs',
+        '.babelrc.cts',
+        '.babelrc',
+      ];
+
+      for (const configFile of babelConfigFiles) {
+        const babelConfigPath = path.join(config.modRequest.projectRoot, configFile);
+
+        if (!fs.existsSync(babelConfigPath)) {
```
Comment (Kudo) [inline packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts:85]: maybe we could get faster by async calls. could you try to apply the change?

```diff
diff --git a/packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts b/packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts
index d4698761ee..a9f9442714 100644
--- a/packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts
+++ b/packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts
@@ -65,15 +65,48 @@ const withCliBabelConfig: ConfigPlugin = (config) => {
   return withDangerousMod(config, [
     'ios',
     async (config) => {
-      const babelConfigPath = path.join(config.modRequest.projectRoot, 'babel.config.js');
-      let contents = await fs.promises.readFile(babelConfigPath, 'utf8');
-      contents = updateBabelConfig(contents);
-      await fs.promises.writeFile(babelConfigPath, contents);
+      try {
+        const babelConfigPath = await findBabelConfigPathAsync(config.modRequest.projectRoot);
+        let contents = await fs.promises.readFile(babelConfigPath, 'utf8');
+        contents = updateBabelConfig(contents);
+        await fs.promises.writeFile(babelConfigPath, contents);
+      } catch {
+        console.warn(
+          '⚠️  Could not find Babel config file in the project root. Please manually update the Babel config to use `babel-preset-expo`.'
+        );
+      }
       return config;
     },
   ]);
 };
 
+async function findBabelConfigPathAsync(projectRoot: string): Promise<string> {
+  const babelConfigFiles = [
+    'babel.config.json',
+    'babel.config.js',
+    'babel.config.cjs',
+    'babel.config.mjs',
+    'babel.config.cts',
+    '.babelrc.json',
+    '.babelrc.js',
+    '.babelrc.cjs',
+    '.babelrc.mjs',
+    '.babelrc.cts',
+    '.babelrc',
+  ];
+
+  return Promise.any(
+    babelConfigFiles.map(async (filename) => {
+      const configPath = path.join(projectRoot, filename);
+      const stat = await fs.promises.stat(configPath);
+      if (!stat.isFile()) {
+        throw new Error(`Config file is not a file: ${configPath}`);
+      }
+      return configPath;
+    })
+  );
+}
+
 const withCliMetroConfig: ConfigPlugin = (config) => {
   return withDangerousMod(config, [
     'ios',
@@ -114,7 +147,10 @@ export function updateAndroidGradleFile(contents: string): string {
 }
 
 export function updateBabelConfig(contents: string): string {
-  return contents.replace(/['"]module:metro-react-native-babel-preset['"]/g, `'babel-preset-expo'`);
+  return contents.replace(
+    /(['"])module:(metro-react-native-babel-preset|@react-native\/babel-preset)(['"])/g,
+    `$1babel-preset-expo$3`
+  );
 }
 
 export function updateMetroConfig(contents: string): string {
```

this diff comes with another fix that the `metro-react-native-babel-preset` is outdated from react-native 0.74.

**Review by Kudo — CHANGES_REQUESTED** [review]
good point for the comment and changes. leaving a comment for all async calls and we should be good to merge

File: packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts (hunk 4 above)
Comment (wojtekmaj) [reply packages/install-expo-modules/src/plugins/cli/withCliIntegration.ts]: Nice one. Applied!

**Comment by Kudo** [comment]
looks like ci was failed from some inconsistent code style with out prettier settings. could you try running `yarn lint --fix` to fix those errors?

**Comment by wojtekmaj** [comment]
Yeah this PR is a pain in the butt for me since I have to do it all from GitHub UI because the repo is too big to clone :D Let's see how this goes this time.

**Comment by Kudo** [comment]
> Yeah this PR is a pain in the butt for me since I have to do it all from GitHub UI because the repo is too big to clone :D Let's see how this goes this time.

totally understand the pain point. ci all passed, here we go 🚀 thanks again for having this pr 🫡

**Review by Kudo — APPROVED** [review]

**Comment by vonovak** [comment]
@wojtekmaj thank you for your PR. btw to make cloning much faster, use a customized git clone command as documented here https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-download-and-setup 🙂

**Comment by wojtekmaj** [comment]
Is there a release schedule somewhere by any chance? Really looking forward to test this feature out in practice!

**Comment by wojtekmaj** [comment]
It's been more than 2 months, can we get this released?

**Comment by vonovak** [comment]
Hello @wojtekmaj,
Thank you for your patience - new features are typically released in the next SDK, which would be Expo SDK 52. You can expect it out in a couple weeks.

Please use local patches in the mean time.
Hope this helps,
Vojtech

**Pull request merged**